File: app/src/main/java/com/veganbeauty/app/features/account/expiry/AccountProductExpiryDetailCircularProgress.java
package com.veganbeauty.app.features.account.expiry;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.Nullable;

public class AccountProductExpiryDetailCircularProgress extends View {

    private float progress = 0.0f;
    private int progressColor = Color.parseColor("#3E4D44");

    private final Paint backgroundPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint progressPaint = new Paint(Paint.ANTI_ALIAS_FLAG);

    public AccountProductExpiryDetailCircularProgress(Context context) {
        super(context);
        init();
    }

    public AccountProductExpiryDetailCircularProgress(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public AccountProductExpiryDetailCircularProgress(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        backgroundPaint.setColor(Color.parseColor("#E0E4E2"));
        backgroundPaint.setStyle(Paint.Style.STROKE);
        backgroundPaint.setStrokeWidth(dp(6f));

        progressPaint.setColor(progressColor);
        progressPaint.setStyle(Paint.Style.STROKE);
        progressPaint.setStrokeWidth(dp(6f));
        progressPaint.setStrokeCap(Paint.Cap.ROUND);
    }

    public void setProgress(float value) {
        this.progress = Math.max(0f, Math.min(1f, value));
        invalidate();
    }

    public float getProgress() {
        return progress;
    }

    public void setProgressColor(int color) {
        this.progressColor = color;
        progressPaint.setColor(color);
        invalidate();
    }

    public int getProgressColor() {
        return progressColor;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        float w = getWidth();
        float h = getHeight();
        float size = Math.min(w, h);
        float stroke = dp(6f);
        float radius = (size - stroke) / 2f;

        float cx = w / 2f;
        float cy = h / 2f;

        canvas.drawCircle(cx, cy, radius, backgroundPaint);

        RectF rectF = new RectF(cx - radius, cy - radius, cx + radius, cy + radius);
        canvas.drawArc(rectF, -90f, progress * 360f, false, progressPaint);
    }

    private float dp(float value) {
        return value * getContext().getResources().getDisplayMetrics().density;
    }
}

File: app/src/main/java/com/veganbeauty/app/features/account/expiry/AccountProductExpiryDetailCircularProgress.kt
package com.veganbeauty.app.features.account.expiry

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View

class AccountProductExpiryDetailCircularProgress @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    // Progress from 0.0 to 1.0
    var progress: Float = 0.0f
        set(value) {
            field = value.coerceIn(0f, 1f)
            invalidate()
        }

    private val backgroundPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.parseColor("#E0E4E2") // Gray background ring
        style = Paint.Style.STROKE
        strokeWidth = dp(6f)
    }

    private val progressPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.parseColor("#3E4D44") // Primary dark green ring
        style = Paint.Style.STROKE
        strokeWidth = dp(6f)
        strokeCap = Paint.Cap.ROUND
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        val w = width.toFloat()
        val h = height.toFloat()
        val size = minOf(w, h)
        val stroke = dp(6f)
        val radius = (size - stroke) / 2f

        val cx = w / 2f
        val cy = h / 2f

        // Draw background circle
        canvas.drawCircle(cx, cy, radius, backgroundPaint)

        // Draw progress arc (start from top -90 degrees)
        val rectF = RectF(cx - radius, cy - radius, cx + radius, cy + radius)
        canvas.drawArc(rectF, -90f, progress * 360f, false, progressPaint)
    }

    private fun dp(value: Float): Float {
        return value * context.resources.displayMetrics.density
    }
}
